Add preview extension for Markdown

AlexTugarev opened the pull request:
The preview extension should provide:
- [x] context menu actions to open preview of `*.md` files ...
  - [x] ... to the side, which depends on #1028
  - [x] ~reuse single instance of preview~
- [x] sync with current editor
  - [x] reveal the part that was generated from the line at cursor position in editor
  - [x] scroll editor to the selected position of generated output 
- [x] ensure styling works with light theme, too
- [x] make Markdown preview a contribution to generic preview extension
  - [x] refactor rename the extension
  - [x] enable previewer contributions
- [x] enable multiple preview widgets
- [x] use relative path's of links to navigate to resources

This extension is based no https://github.com/theia-ide/theia-extension-example.

Quick demo:
![2018-01-23 12 31 35](https://user-images.githubusercontent.com/914497/35273685-bf4f678e-0039-11e8-9232-eab7071af625.gif)

AlexTugarev commented:
Currently, this extension is coined to be a preview for Markdown only. A look at the Markdown specifics shows, that it's possible to turn it into a common preview extension, which uses contributions to support different resource types. Such preview contributions should contain `ResourceResolver` and a(n optional) strategy to reveals current cursor position.

AlexTugarev commented:
@svenefftinge' comment:
> 1. scrolling. It should scroll the editor, when the preview is scrolled.

Sure. I think `TextEditor.revealPosition` will do that.

> 2. I think it would be good to make this a singleton widget, that is reused and repopulated whenever a 'review-enabled' editor gets focus. We could add it to the right panel on first open.

We agreed on a single preview widget, but leave it in main panel.

I've updated the TODO list from above.

AlexTugarev commented:
@akosyakov is there a reason, why `MonacoEditor` translates `revealPosition` to `revealPositionInCenter`? cf. https://github.com/theia-ide/theia/blob/master/packages/monaco/src/browser/monaco-editor.ts#L186

For scroll synchronization it would be nice to use the actual `revealPosition` operation.

akosyakov commented:
@AlexTugarev existing clients expecting that `revealPosition` reveals in the center, like for navigation between editors. But go ahead rename existing method and a new one.

akosyakov commented on packages/preview/src/browser/markdown/markdown-preview-handler.ts:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    @inject(Workspace)
+    protected readonly workspace: Workspace;
```
it does not seem to be used

it would be nice if unused code which is not present in master won't be removed by separate commits, you can fix such changes up to the original commit

AlexTugarev replied:
I'd like to squash all children of commit https://github.com/theia-ide/theia/pull/1051/commits/911d7735a3df631ab3db8b6a82d3083df8d3bd87 in the end. Would that be okay?

akosyakov replied:
it is fine, i usually do interactive rebase immediately to preserve logical order of commits locally

akosyakov commented on packages/preview/src/browser/preview-contribution.ts:
```diff
@@ -0,0 +1,163 @@
+    readonly label = 'Preview';
+
+    protected readonly disposables = new DisposableCollection();
+    protected previewWidget: PreviewWidget | undefined;
```
it should not be necessary, you can always get it from the widget manager, right?

AlexTugarev replied:
sounds good.

akosyakov commented on packages/preview/src/browser/preview-handler.ts:
```diff
@@ -0,0 +1,38 @@
+export const PreviewHandler = Symbol('PreviewHandler');
+
+export interface PreviewHandler {
+    canHandle(uri: URI): boolean;
```
Should not it be number based, similar to `OpenHandler` to prioritize several previewers for the same URI?
You can use `Prioritizeable.prioritizeAllSync` to look up a handler, like here: https://github.com/theia-ide/theia/blob/e6902835cfa83895b81b8417c0b5de8731280761/packages/core/src/browser/label-provider.ts#L113-L118

AlexTugarev replied:
sounds good!

akosyakov commented on packages/preview/src/browser/preview-handler.ts:
```diff
@@ -0,0 +1,38 @@
+
+export interface PreviewHandler {
+    canHandle(uri: URI): boolean;
+    renderHTML(content: string): Promise<string | undefined>;
```
please use `MaybePromise` in APIs that extenders don't need to wrap results in promises if not needed

akosyakov commented on packages/preview/src/browser/preview-handler.ts:
```diff
@@ -0,0 +1,38 @@
+/*
```
it would be nice to reexport APIs via `browser/index.ts`

AlexTugarev replied:
it's already reexported.

akosyakov commented on packages/preview/src/browser/preview-contribution.ts:
```diff
@@ -0,0 +1,163 @@
+            previewWidget.disposed.connect(() => {
+                this.previewWidget = undefined;
+            });
+            this.app.shell.addToMainArea(previewWidget, options);
```
you can check `widget.isAttached` to get rid of `this.previewWiwdget`

akosyakov commented on packages/preview/src/browser/preview-widget.ts:
```diff
@@ -0,0 +1,195 @@
+        if (resource.onDidChangeContents) {
+            this.resourceDisposibles.push(resource.onDidChangeContents(() => this.update()));
+        }
+        const updateIfAffected = (affectedUri?: string) => {
```
protected method would be nicer for extenders

akosyakov commented on packages/preview/src/browser/preview-widget.ts:
```diff
@@ -0,0 +1,195 @@
+        }
+    }
+
+    didScrollToLine: ((line: number) => void) | undefined;
```
Is it exposed to clients?

AlexTugarev replied:
shouldn't be.

akosyakov replied:
default modifier is public

akosyakov commented on packages/preview/src/browser/preview-widget.ts:
```diff
@@ -0,0 +1,195 @@
+
+    onActivateRequest(msg: Message): void {
+        super.onActivateRequest(msg);
+        this.update();
```
closeable widgets should focus something on the activation

AlexTugarev replied:
not sure about this. the scenario is: the user is opening the preview to the side of current editor, and continues to modify source in editor. blurring editor would mean, that the user needs to select the editor again.

akosyakov replied:
ok, makes sense

svenefftinge replied:
@spoenemann does that cause any trouble, if the widget doesn't get tracked?

AlexTugarev replied:
Thanks @spoenemann! I've fixed this.

akosyakov commented on packages/preview/src/browser/preview-widget.ts:
```diff
@@ -0,0 +1,195 @@
+    }
+
+    didScrollToLine: ((line: number) => void) | undefined;
+    addDidScrollToLineHandler(handler: (line: number) => void): Disposable {
```
we should use events for such cases

akosyakov commented on packages/preview/src/browser/style/markdown.css:
```diff
@@ -0,0 +1,251 @@
+/*---------------------------------------------------------------------------------------------
```
please add our copyright as well

AlexTugarev replied:
actually, I'm not sure, should we leave the theming switch based on `.vs` and `.vs-dark` as is, or rewrite with `theia`?

akosyakov replied:
theia based will be better if it is possible

akosyakov commented on packages/preview/src/browser/style/tomorrow.css:
```diff
@@ -0,0 +1,73 @@
+/* Tomorrow Theme */
```
please add our copyright as well

akosyakov commented on packages/preview/src/package.spec.ts:
```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright (C) 2017 TypeFox and others.
```
2018

AlexTugarev commented:
For synchronizing scrolling of preview with editor, it would be nice to have `TextEditor.revealPosition` to not center horizontally. That would require to introduce `TextEditor.revealPositionInCenter` and change the implementation `MonacoEditor.revealPosition`, which would affect `TextEditor` clients and require adaptions.

AlexTugarev commented:
Preview contributions should also provide styling.

AlexTugarev commented:
- [ ] issue: reveal of section in preview should not trigger reveal in editor

svenefftinge commented:
Somehow the syncing is one change behind:
![sync-issue](https://user-images.githubusercontent.com/372735/35051670-13381e42-fba6-11e7-9ca6-b43154aa11d2.gif)

svenefftinge commented:
On first opening the preview, the view's scrollposition should sync with the editor's selection.

svenefftinge commented:
The scrolling in the preview doesn't really sync the editor, well. It is very slow and choppy for me.

AlexTugarev commented:
@svenefftinge :
> Somehow the syncing is one change behind:

It depends on the DidChangeTextDocument cycle. Have a look at the dirty state indicator of the editor.

svenefftinge commented:
> It depends on the DidChangeTextDocument cycle. Have a look at the dirty state indicator of the editor.
No, it is one change behind for me. I can wait for minutes and will not see the current state of the document reflected in the preview, but only the previous one.

AlexTugarev commented:
@svenefftinge : 
> The scrolling in the preview doesn't really sync the editor, well. It is very slow and choppy for me.

I've identified two related issues:
- [ ] reveal in editor should not center horizontally in this case
- [ ] more levels than just top one should be marked with line position to make reveal more accurate

AlexTugarev commented:
@svenefftinge :
> No, it is one change behind for me. I can wait for minutes and will not see the current state of the document reflected in the preview, but only the previous one.

Sorry, it was hard to see in the low-res gif from above. Not sure, how to reproduce it. I'll try to test reproduce it somehow. This is, how it is for me at the moment:

![2018-01-17 17 12 07](https://user-images.githubusercontent.com/914497/35053206-c1890de6-fba9-11e7-95d3-da03c9335181.gif)

svenefftinge commented:
I was able to reproduce after closing and reopening the editor and view. We can have a look together via hangout tomorrow.

AlexTugarev commented:
@svenefftinge, besides the "one change behind" issue, I've fixed the delayed update of the preview. basically it was always reading contents from resource, now it reads from text document, too.

![2018-01-18 06 43 24](https://user-images.githubusercontent.com/914497/35082606-23302180-fc1b-11e7-923f-66fb687a0403.gif)

svenefftinge commented on packages/preview/src/browser/preview-widget.ts:
```diff
@@ -0,0 +1,253 @@
+
+    protected preventScrollNotification: boolean = false;
+    protected startScrollSync(): void {
+        this.node.addEventListener('scroll', throttle(50, (event: UIEvent) => {
```
How is `throttle` different to `setTimeout`?

AlexTugarev replied:
`throttle` reduces the frequency of events here. OTOH the previous delay strategy with `setTimeout`/`clearTimeout` caused lagging of sync.

AlexTugarev commented on packages/preview/src/browser/preview-handler.ts:
```diff
@@ -0,0 +1,42 @@
+    readonly contentClass: string;
+    canHandle(uri: URI): number;
+    renderHTML(content: string): MaybePromise<string | undefined>;
+    findElementForSourceLine(sourceLine: number, renderedNode: HTMLElement): HTMLElement | undefined;
```
good idea!

AlexTugarev commented on packages/preview/src/browser/preview-contribution.ts:
```diff
@@ -0,0 +1,194 @@
+
+    protected getPreviewWidget(): PreviewWidget | undefined {
+        // note: this also ensures, we get resotored widgets
+        const previewWidget = this.widgetManager.getWidgets(PREVIEW_WIDGET_FACTORY_ID).slice(-1)[0] as PreviewWidget;
```
though, this will be obsolete with the "multi widget" scenario, the original intention to use `WidgetManager.getWidgets` was to register event listeners for the instance, that was restored after start. I need to admin, that I switched to a widget factory which can now notify about created widgets, but I left this `getPreviewWidget`.

also, I did not yet apply Anton's hint to test, if widget is attached, in order to get finally rid of `WidgetManager.getWidgets` calls.

AlexTugarev commented on packages/preview/src/browser/markdown/markdown-preview-handler.ts:
```diff
@@ -0,0 +1,109 @@
+    protected engine: markdownit.MarkdownIt | undefined;
+    protected getEngine(): markdownit.MarkdownIt {
+        if (!this.engine) {
+            const engine: markdownit.MarkdownIt = this.engine = markdownit({
```
yes, I also like the idea of renderer extensions. 'sounds familiar to me ;-)

akosyakov replied:
It would be nice to decouple the markdown renderer to be able to use it outside of the previewer.

AlexTugarev replied:
@akosyakov, there are several options how to extract the markdownit engine from this preview handler. it may depend on other requirements. 

configuring of plugins and renderers happens at construction time. while additional renderers could be controlled through render method's options parameter, I see no way to disable plugins. OTOH we could have a factory for markdownit engines with options, e.g. `createMarkdownEngine(options: MarkdownEngineOptions) : MarkdownIt`

svenefftinge replied:
Let's do that in another step.

AlexTugarev commented:
> It would be double good, if the relative paths would render links that open the respective editor / preview.
> Also, the singleton preview widget would only make sense if it would automatically sync on editor focus change. But in light of navigating around I changed my mind about this, and think multiple preview widgets would work better 

I've updated the description from above.

spoenemann commented:
> the scenario is: the user is opening the preview to the side of current editor, and continues to modify source in editor. blurring editor would mean, that the user needs to select the editor again.

If you don't want the preview to steal the focus, then don't activate it. You can add widgets to the shell without activating them.

> does that cause any trouble, if the widget doesn't get tracked?

Yes. See #1061.

spoenemann commented:
I created #1093 for the focus vs. activation issue.

AlexTugarev commented:
@svenefftinge, sorry NVM, I think I just forgotten to make the widget id unique again.

AlexTugarev commented:
@svenefftinge, could you give it another try, please!

AlexTugarev commented on packages/preview/src/browser/markdown/style/markdown.css:
```diff
@@ -0,0 +1,162 @@
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+/*---------------------------------------------------------------------------------------------
```
@svenefftinge and @akosyakov, I've changed the most of markdown.css, which includes:
- theming 
- highlighting elements with line info
- fixed table styles
- removed .vs(-dark)

Should we still have this (c) block?

svenefftinge replied:
Could you share a link to the original?

svenefftinge replied:
Is it this one? https://github.com/Microsoft/vscode/blob/master/extensions/markdown/media/markdown.css
Looks very different.

svenefftinge replied:
But still, I think you could keep it and the eclipse IP team can check that again.

AlexTugarev replied:
Sorry, should be already attached. I copied from https://github.com/theia-ide/theia-extension-example/blob/master/theia-markdown/src/browser/style/markdown.css, which is a copy of 
https://github.com/Microsoft/vscode/blob/master/extensions/markdown/media/markdown.css

AlexTugarev replied:
That's okay.

svenefftinge commented on packages/preview/README.md:
```diff
@@ -0,0 +1,8 @@
+# Theia - Preview Extension
+
+It comes with built-in support for rendering markdown files.
```
Sorry, I did not mean to say replace your sentence, but add to it :-)

AlexTugarev replied:
✔️

svenefftinge commented on packages/preview/src/browser/preview-widget.ts:
```diff
@@ -0,0 +1,278 @@
+        this.previewDisposables.dispose();
+    }
+
+    async start(uri: URI): Promise<void> {
```
`updateContent` would be a better name.

AlexTugarev replied:
✔️

AlexTugarev commented:
@svenefftinge I changed the color of inlined `code` instead:

<img width="171" alt="screen shot 2018-01-25 at 11 22 07" src="https://user-images.githubusercontent.com/914497/35383856-d0eb7088-01c3-11e8-96f8-0c810d4f1342.png">
<img width="158" alt="screen shot 2018-01-25 at 11 21 47" src="https://user-images.githubusercontent.com/914497/35383866-db45cd44-01c3-11e8-8042-8c511f852abf.png">

what do you think?

AlexTugarev commented:
click on links with `relative/path/to/NotExisting.md` opens empty preview widget. 

what should happen instead? show a warning?

svenefftinge commented:
Looks really good now. Still I have comments :-)

1) When I open a *.md in preview and in there click a link to another *.md it opens in preview. When I then double-click on the second preview, the opened isn't synced, i.e. changes are note reflected in the preview.

2) In the Shell, the opening to the side, should only do a splitting, if there not already is a split that can be reused.

3) Double clicking in a markdown should open the editor to the side, if it has to open an editor.

svenefftinge commented:
Please squash the commits to a sensible amount of units.

svenefftinge commented:
> click on links with relative/path/to/NotExisting.md opens empty preview widget.
> what should happen instead? show a warning?

Tell the user through notifications.

AlexTugarev commented:
@svenefftinge I've tackled first two issue, but for 
> Double clicking in a markdown should open the editor to the side, if it has to open an editor.

we'd need to touch `EditorManager.open` and add `EditorOpenerOptions ` like this:
```
interface EditorOpenerOptions extends EditorInput | ApplicationShell.WidgetOptions { }
```
Otherwise we cannot control where to insert the editor.

svenefftinge commented on packages/editor/src/browser/editor-manager.ts:
```diff
@@ -90,17 +94,21 @@ export class EditorManagerImpl implements EditorManager, WidgetFactory {
-    canHandle(uri: URI, input?: EditorInput): number {
-        return 100;
+    canHandle(uri: URI, options?: EditorOpenerOptions): number {
+        return uri.scheme === 'file' ? 100 : 0;
```
sorry, my fault. But it should actually not restrict to `file` protocol. E.g. `git` URIs are opened through this as well. 
It should handle everything for which a `DefaultResourceProvider` returns a `Resource`. Please just revert it to always returning `100`.

svenefftinge commented on packages/core/src/browser/http-open-handler.ts:
```diff
@@ -0,0 +1,25 @@
+        return uri.scheme.startsWith('http') ? 500 : 0;
+    }
+
+    async open(uri: URI) {
```
superfluous `async`

svenefftinge commented on packages/editor/src/browser/editor-manager.ts:
```diff
@@ -90,17 +94,21 @@ export class EditorManagerImpl implements EditorManager, WidgetFactory {
-                this.app.shell.addWidget(editor, { area: 'main' });
+                const widgetOptions: ApplicationShell.WidgetOptions = (options && options.widgetOptions)
+                    ? options.widgetOptions
+                    : { area: 'main' };
```
Not important, but `options && options.widgetOptions || { area: 'main' }` should work, too.

svenefftinge commented on packages/preview/src/browser/preview-widget-manager.ts:
```diff
@@ -0,0 +1,60 @@
+    protected readonly onWidgetCreatedEmitter = new Emitter<string>();
+
+    protected readonly disposables = new DisposableCollection();
+    private widgets = new Map<string, PreviewWidget>();
```
you don't need to maintain your own cache. The `WidgetManager`does that already.

AlexTugarev replied:
the main purpose of `PreviewWidgetManager` is to provide `onWidgetCreated` event, which is used to update synchronization with a maybe existing editor widget. 

unfortunately the `WidgetManager` will add that widget a step after calling `createWidget` on the factory. but it's possible to refactor this, and to notify with the created widget instead of uri.

svenefftinge commented on packages/preview/src/browser/preview-widget-manager.ts:
```diff
@@ -0,0 +1,60 @@
+        if (previewWidget) {
+            return previewWidget;
+        }
+        const newWidget = this.container.get(PreviewWidget);
```
Here the `uri` should be added to the widget (ideally through constructor).

AlexTugarev replied:
done

svenefftinge commented on packages/preview/src/browser/preview-widget.ts:
```diff
@@ -0,0 +1,258 @@
+        return this.previewHandler.renderContent({ content, originUri });
+    }
+
+    storeState(): object {
```
The uri is the key for the widget and should be set by the factory. So there is no need for storing it here again.
The store/restore is really just to store state that is heavy to compute (e.g. error markers, git changes, etc.) In the case of preview, it would be the result from the PreviewHandler. But my feeling is that it would be quite heavy to store it in the browser storage and it is not so hard to recompute, so we better just don't implement `StatefulWidget`.

AlexTugarev replied:
Thanks for pointing this out. I've cleaned that up. I've did this similar to the `TerminalWidget` when the preview widget meant to be a singleton. 
Actually, I was not aware that, the `createWidget`'s options are persisted, that's cool!

akosyakov commented on packages/preview/src/common/test/jsdom.ts:
```diff
@@ -0,0 +1,23 @@
+
+import { JSDOM } from 'jsdom';
+
+const dom = new JSDOM('<!doctype html><html><body></body></html>');
```
Should not it happen before each test and after each test global should be cleaned up?

AlexTugarev replied:
Thanks! I forgot about the cleanup. It's now similar to `global-jsdom` and cleanup is done in `after`, cf. https://github.com/theia-ide/theia/blob/e2bfb21b98fa714cbd1b5558ae11d39bbf4d21cb/packages/preview/src/browser/markdown/markdown-preview-handler.spec.ts#L24

Also, it's important to execute this jsdom helper before the modules are loaded which rely on certain window properties. This is why it's not working to call it in `before`.

epatpol replied:
Ah so that might be why my tests were failing [here](https://gist.github.com/epatpol/6ede50cf5da7d08ba2050558c6880a7f) even though I used jsdom() in the `before` mocha hook?

akosyakov replied:
@epatpol please pull this code up to @theia/core/src/browser/test and move jsdom dependency to the root package

epatpol replied:
Ok will do!

svenefftinge commented on packages/core/src/browser/frontend-application-module.ts:
```diff
@@ -53,6 +54,8 @@ export const frontendApplicationModule = new ContainerModule((bind, unbind, isBo
     bind(DefaultOpenerService).toSelf().inSingletonScope();
     bind(OpenerService).toDynamicValue(context => context.container.get(DefaultOpenerService));
+    bind(HttpOpenHandler).toSelf().inSingletonScope();
+    bind(OpenHandler).toDynamicValue(ctx => ctx.container.get(HttpOpenHandler)).inSingletonScope();
```
Just a remark  and not a good enough reason for another round :)
If there is no client for `HttpOpenHandler` specifically, a simple binding like below is sufficient:
```
bind(OpenHandler).to(HttpOpenHandler).inSingletonScope();
```
